### PR TITLE
[HAL] Fix more bugs in HalpDebugPciDumpBus

### PR DESCRIPTION
**Pull request by binarymaster**
## Purpose

Fix more bugs in HalpDebugPciDumpBus.

JIRA issue: N/A

## Proposed changes

- Don't match device id outside current vendor id list.
- Don't match subsystem id outside current device id list.
- Refactor some magic constants.

Addendum to 894635b.

#### BEFORE
```
00:01.3 Bridge [0680]: MediaQ Inc. MQ-200 [4d51:0200] (rev 03)
	Subsystem: Unknown [0000:0000]
00:01.3 Bridge [0680]: MediaQ Inc. [mvHYPERION-CLe/CLb] CameraLink PCI Express x1 Frame Grabber [4d51:0000] (rev 03)
	Subsystem: Unknown [0000:0000]
00:01.3 Bridge [0680]: MediaQ Inc. [mvHYPERION-CLe/CLb] CameraLink PCI Express x1 Frame Grabber [4d51:0000] (rev 03)
	Subsystem: 86C775 Trio64V2/DX [5333:8900]
00:01.3 Bridge [0680]: S3 Graphics Ltd. 86c755 [Trio 64V2/DX] [5333:8900] (rev 03)
	Subsystem: 86C775 Trio64V2/DX [5333:8900]
00:01.3 Bridge [0680]: S3 Graphics Ltd. 86c755 [Trio 64V2/DX] [5333:8900] (rev 03)
	Subsystem: 86C775 Trio64V2/DX, 86C785 Trio64V2/GX [5333:8901]
00:01.3 Bridge [0680]: S3 Graphics Ltd. 86c775/86c785 [Trio 64V2/DX or /GX] [5333:8901] (rev 03)
	Subsystem: 86C775 Trio64V2/DX, 86C785 Trio64V2/GX [5333:8901]
00:01.3 Bridge [0680]: NVIDIA Corporation NV2A [XGPU] [10de:02a0] (rev 03)
	Subsystem: GLoria L [0000:0000]
00:01.3 Bridge [0680]: 3DLabs GLINT 500TX [3d3d:0002] (rev 03)
	Subsystem: GLoria L [0000:0000]
00:01.3 Bridge [0680]: 3DLabs GLINT 500TX [3d3d:0002] (rev 03)
	Subsystem: GLoria XXL [1048:0a42]
00:01.3 Bridge [0680]: 3DLabs Permedia [3d3d:0004] (rev 03)
	Subsystem: GLoria XL [0000:0000]
00:01.3 Bridge [0680]: 3DLabs Permedia [3d3d:0005] (rev 03)
	Subsystem: GLoria XL [0000:0000]
00:01.3 Bridge [0680]: 3DLabs GLINT MX [3d3d:0006] (rev 03)
	Subsystem: GLoria XL [0000:0000]
00:01.3 Bridge [0680]: 3DLabs GLINT MX [3d3d:0006] (rev 03)
	Subsystem: GLoria XXL [1048:0a42]
```
#### AFTER
```
00:01.3 Bridge [0680]: MediaQ Inc. MQ-200 [4d51:0200] (rev 03)
	Subsystem: Unknown [0000:0000]
00:01.3 Bridge [0680]: MediaQ Inc. Unknown device [4d51:0000] (rev 03)
	Subsystem: Unknown [0000:0000]
00:01.3 Bridge [0680]: MediaQ Inc. Unknown device [4d51:0000] (rev 03)
	Subsystem: Unknown [5333:8900]
00:01.3 Bridge [0680]: S3 Graphics Ltd. 86c755 [Trio 64V2/DX] [5333:8900] (rev 03)
	Subsystem: 86C775 Trio64V2/DX [5333:8900]
00:01.3 Bridge [0680]: S3 Graphics Ltd. 86c755 [Trio 64V2/DX] [5333:8900] (rev 03)
	Subsystem: Unknown [5333:8901]
00:01.3 Bridge [0680]: S3 Graphics Ltd. 86c775/86c785 [Trio 64V2/DX or /GX] [5333:8901] (rev 03)
	Subsystem: 86C775 Trio64V2/DX, 86C785 Trio64V2/GX [5333:8901]
00:01.3 Bridge [0680]: NVIDIA Corporation NV2A [XGPU] [10de:02a0] (rev 03)
	Subsystem: Unknown [0000:0000]
00:01.3 Bridge [0680]: 3DLabs GLINT 500TX [3d3d:0002] (rev 03)
	Subsystem: GLoria L [0000:0000]
00:01.3 Bridge [0680]: 3DLabs GLINT 500TX [3d3d:0002] (rev 03)
	Subsystem: Unknown [1048:0a42]
00:01.3 Bridge [0680]: 3DLabs Permedia [3d3d:0004] (rev 03)
	Subsystem: Unknown [0000:0000]
00:01.3 Bridge [0680]: 3DLabs Permedia [3d3d:0005] (rev 03)
	Subsystem: Unknown [0000:0000]
00:01.3 Bridge [0680]: 3DLabs GLINT MX [3d3d:0006] (rev 03)
	Subsystem: GLoria XL [0000:0000]
00:01.3 Bridge [0680]: 3DLabs GLINT MX [3d3d:0006] (rev 03)
	Subsystem: GLoria XXL [1048:0a42]
```
#### DIFF
```diff
 00:01.3 Bridge [0680]: MediaQ Inc. MQ-200 [4d51:0200] (rev 03)
 	Subsystem: Unknown [0000:0000]
-00:01.3 Bridge [0680]: MediaQ Inc. [mvHYPERION-CLe/CLb] CameraLink PCI Express x1 Frame Grabber [4d51:0000] (rev 03)
+00:01.3 Bridge [0680]: MediaQ Inc. Unknown device [4d51:0000] (rev 03)
 	Subsystem: Unknown [0000:0000]
-00:01.3 Bridge [0680]: MediaQ Inc. [mvHYPERION-CLe/CLb] CameraLink PCI Express x1 Frame Grabber [4d51:0000] (rev 03)
-	Subsystem: 86C775 Trio64V2/DX [5333:8900]
+00:01.3 Bridge [0680]: MediaQ Inc. Unknown device [4d51:0000] (rev 03)
+	Subsystem: Unknown [5333:8900]
 00:01.3 Bridge [0680]: S3 Graphics Ltd. 86c755 [Trio 64V2/DX] [5333:8900] (rev 03)
 	Subsystem: 86C775 Trio64V2/DX [5333:8900]
 00:01.3 Bridge [0680]: S3 Graphics Ltd. 86c755 [Trio 64V2/DX] [5333:8900] (rev 03)
-	Subsystem: 86C775 Trio64V2/DX, 86C785 Trio64V2/GX [5333:8901]
+	Subsystem: Unknown [5333:8901]
 00:01.3 Bridge [0680]: S3 Graphics Ltd. 86c775/86c785 [Trio 64V2/DX or /GX] [5333:8901] (rev 03)
 	Subsystem: 86C775 Trio64V2/DX, 86C785 Trio64V2/GX [5333:8901]
 00:01.3 Bridge [0680]: NVIDIA Corporation NV2A [XGPU] [10de:02a0] (rev 03)
-	Subsystem: GLoria L [0000:0000]
+	Subsystem: Unknown [0000:0000]
 00:01.3 Bridge [0680]: 3DLabs GLINT 500TX [3d3d:0002] (rev 03)
 	Subsystem: GLoria L [0000:0000]
 00:01.3 Bridge [0680]: 3DLabs GLINT 500TX [3d3d:0002] (rev 03)
-	Subsystem: GLoria XXL [1048:0a42]
+	Subsystem: Unknown [1048:0a42]
 00:01.3 Bridge [0680]: 3DLabs Permedia [3d3d:0004] (rev 03)
-	Subsystem: GLoria XL [0000:0000]
+	Subsystem: Unknown [0000:0000]
 00:01.3 Bridge [0680]: 3DLabs Permedia [3d3d:0005] (rev 03)
-	Subsystem: GLoria XL [0000:0000]
+	Subsystem: Unknown [0000:0000]
 00:01.3 Bridge [0680]: 3DLabs GLINT MX [3d3d:0006] (rev 03)
 	Subsystem: GLoria XL [0000:0000]
 00:01.3 Bridge [0680]: 3DLabs GLINT MX [3d3d:0006] (rev 03)
 	Subsystem: GLoria XXL [1048:0a42]
```

### CODE REVIEW

File: hal/halx86/legacy/bussupp.c [hunk 1]
```diff
@@ -854,36 +854,58 @@ HalpDebugPciDumpBus(IN ULONG i,
...
+        VendorName += strlen("\r\n0000  ");
         p = strpbrk(VendorName, "\r\n");
         Length = p - VendorName;
         if (Length >= sizeof(bVendorName)) Length = sizeof(bVendorName) - 1;
         strncpy(bVendorName, VendorName, Length);
         bVendorName[Length] = '\0';
+        p += strlen("\r\n");
+        while (*p == '\t' || *p == '#')
+        {
+            p = strpbrk(p, "\r\n");
```
Comment (HBelusca) [inline hal/halx86/legacy/bussupp.c]: Is it possible that strpbrk returns NULL in this condition?

File: hal/halx86/legacy/bussupp.c [hunk 2]
```diff
@@ -854,36 +854,58 @@ HalpDebugPciDumpBus(IN ULONG i,
...
+            ProductName += strlen("\t0000  ");
             p = strpbrk(ProductName, "\r\n");
             Length = p - ProductName;
             if (Length >= sizeof(bProductName)) Length = sizeof(bProductName) - 1;
             strncpy(bProductName, ProductName, Length);
             bProductName[Length] = '\0';
+            p += strlen("\r\n");
+            while ((*p == '\t' && *(p + 1) == '\t') || *p == '#')
+            {
+                p = strpbrk(p, "\r\n");
```
Comment (HBelusca) [inline hal/halx86/legacy/bussupp.c]: Is it possible that strpbrk returns NULL in this condition?

File: hal/halx86/legacy/bussupp.c (hunk 1 above)
Comment (binarymaster) [reply hal/halx86/legacy/bussupp.c]: It's only possible if the PCI database file is broken. :slightly_smiling_face: In details:

If condition `if (VendorName)` is true, we found a line in file. Each line ends with `\r\n`. Let's assume we are at the last line. `p += strlen("\r\n");` will make variable point to the end of file, in our case entire file is a C string, so `*p` will be equal to 0, and loop condition `while (*p == '\t' || *p == '#')` will evaluate to false, thus `strpbrk` is not executed.

File: hal/halx86/legacy/bussupp.c (hunk 2 above)
Comment (binarymaster) [reply hal/halx86/legacy/bussupp.c]: Same for the second loop.

File: hal/halx86/legacy/bussupp.c (hunk 1 above)
Comment (HBelusca) [reply hal/halx86/legacy/bussupp.c]: ok so we expect that there is a trailing newline in the file.

**Pull request merged**